### ISSUE TITLE
imported types remain in compiled code after enabling this plugin

### ISSUE THREAD

**Issue by AndyOGo** [banyudu/ts-import-rename-plugin#2]
I use this plugin to rename CSS imports with this config:

```json
{ "transform": "ts-rename-import-plugin", "type": "config", "from": "/index.css$", "to": "/index.bundle.css" }
```

I `import` some types in my TS file.

```ts
import { SomeType } from "./types.ts";
```

With this plugin the `import/require` is not removed anymore from the compiled file, though the access of the type name is gone.

```ts
const types = require("./types.ts");
```

**Comment by AndyOGo** [comment]
What if you would check, that the transform actually changes the `import`, before altering the node.

https://github.com/banyudu/ts-rename-import-plugin/blob/f713b0257104da207175dbe2c0c6854e10c6fee0/src/index.ts#L35-L45

**Comment by AndyOGo** [comment]
I think I found the real issue.

https://github.com/microsoft/TypeScript/issues/40603

**Comment by nonara** [comment]
That's the issue. After a lot of digging and some helpful direction from Andrew Branch @ MSFT, I was able to replicate the TS compiler's elision behaviour. Feel free to copy.

This is the file doing the work:
https://github.com/LeDDGroup/typescript-transform-paths/blob/master/src/utils/elide-import-export.ts

You can see how it's implemented here:
https://github.com/LeDDGroup/typescript-transform-paths/blob/master/src/visitor.ts#L103-L137

**Comment by AndyOGo** [comment]
@banyudu what do think?

**Comment by banyudu** [comment]
> @banyudu what do think?

Thanks, @AndyOGo @nonara ，will fix it soon.

**Comment by AndyOGo** [comment]
@banyudu Thanks for merging my PR.
It handles the case for my issue well.

@nonara approach looks to be the accurate solution for any case.

**Comment by banyudu** [comment]
> @banyudu Thanks for merging my PR.
> It handles the case for my issue well.
> 
> @nonara approach looks to be the accurate solution for any case.

I'm not sure whether it's ok to automatically remove type imports for users. IMO, it seems not a behavior users would expect `ts-rename-import-plugin` to do.

**Comment by AndyOGo** [comment]
Thanks for your quick answer.

I agree with you, this should be properly handled by typescript itself.

This is the default behavior of typescript, that it removes type information and type imports in the output JS (since types aren't valid ECMAScript, they need to be removed, else JS engines would throw syntax errors).

Anyway typescript `updateImportDeclaration` has a bug, as soon as it is used it stops doing the elision process of type imports for the updated `import` and this broke my build - all type imports where still there, without the actual file.

All details can be found in this issue:
https://github.com/microsoft/TypeScript/issues/40603

**Comment by banyudu** [comment]
@AndyOGo Typescript stops doing the elision process of type imports by its design, it's not recognized as a bug yet.

Since `updateImportDeclaration` stops doing the elision process of type imports, it's reasonable to do the elision process by this plugin.

But if this plugin removes the type imports, it may break the plugins after it(if any).

So I think a better way is using the `onion model`:

- rename import paths
- run other plugins
- remove the changed "type imports"

But I don't know how to do it yet.

**Comment by AndyOGo** [comment]
Indeed, I have read that they choose this by design. Though the output clearly shows that it produces faulty results.
Frankly, ts is awesome, though this design decision was just wrong.

You are right, I see your point.
I would be keen what is @nonara experience with possibly breaking other plugins by this change?

**Comment by banyudu** [comment]
Maybe we can just check "type imports" before rename, and skip them.

**Comment by AndyOGo** [comment]
To skip those type imports is also a nice idea. Though how would you differentiate them?
I have also use case where both types and modules are imported from the same ts file.

**Comment by banyudu** [comment]
> To skip those type imports is also a nice idea. Though how would you differentiate them?
> I have also use case where both types and modules are imported from the same ts file.

Emm, you are right.

**Comment by nonara** [comment]
> I would be keen what is @nonara experience with possibly breaking other plugins by this change?

This is a big issue that makes the whole world of transformers quite complex. Chaining transformers will inevitably have interesting side-effects. A well-built transformer will make use of `original` as needed, to walk up the chain in order to get information it needs. 

IMO, with regard to this case, the worst case scenario is having a non-existent identifier imported or exported in your output JS. This causes the entire code to fail. This is both critical and tends to happen often. In contrast, causing side-effects in a subsequent transformer is a much more rare. Additionally, I believe that it becomes even more rare in this case. I'll explain why...

What we've done is we only apply our elision behaviour if the transformer is running as a **JS before** transform. So we don't run it during declarations transform. This means that there will be no usage of any type identifier anywhere in the JS file, so there are very few conceivable reasons that a _JS transformer_ would have any reason to actually need to see unused identifiers in question in the import/export declaration.

However - if it does want to for some reason, it's still able to do this by accessing the node's `original` property to walk up the chain. This is how that scenario should be handled properly, anyway, if it wants itself to be compatible with other transformers.

So to sum it up, I would say that the chances of you causing conflict with another transformer are extremely rare. If they do occur, however, they can be easily remedied by a PR to the other transformer to use `original`. On the other hand, the potential for breaking code in the event of hitting this issue by having elision break down is much more common and critical. That has been our experience anyway!

Hope that helps!

**Comment by nonara** [comment]
It just occurred to me - would changing to an `after` transformer suit this library's use case?

**Comment by banyudu** [comment]
> It just occurred to me - would changing to an `after` transformer suit this library's use case?

In my opinion, it's not a good idea to handle this issue in every transformers. Providing an `after-transform` transformer as a patch to the latest typescript is better.

In this way, users could fix this issue by appending this `after-transform` plugin after the original transformers like `ts-rename-import-plugin` or `typescript-transform-paths`.

**Comment by banyudu** [comment]
How about creating a new `after-transform` transformer from your code? @nonara

**Comment by nonara** [comment]
> How about creating a new `after-transform` transformer from your code? @nonara

It looks like the TS team has started looking into the issue, so there may not be a need to work around it much longer. If it turns out that it doesnt get fixed upstream, I'll make a helper npm package that we can use.

What I mean by `after` is to have your users specify `after` in the transform config. That way it would be transforming the module specifier after type elision has already completed.

That was also just suggested as a remedy by the TS team member looking into the issue. For now, I think I'm going to hold off and see what happens on the TS side.
